components/user-mgt/org.wso2.carbon.identity.user.pre.update.password.action/src/main/java/org/wso2/carbon/identity/user/pre/update/password/action/execution/PreUpdatePasswordActionRequestBuilder.java

```diff
@@ -0,0 +1,219 @@
+    }
+
+    private static Tenant getTenant() {
+
+        return new Tenant(String.valueOf(IdentityContext.getThreadLocalCarbonContext().getTenantId()),
+                IdentityContext.getThreadLocalCarbonContext().getTenantDomain());
+    }
+
+    private User getUser() throws ActionExecutionRequestBuilderException {
+
+        PasswordUpdatingUser.Builder userBuilder = new PasswordUpdatingUser.Builder()
+                .id(userActionContext.getUserId());
+        populateCredential(userBuilder);
+        return userBuilder.build();
+    }
+
+    private void populateCredential(PasswordUpdatingUser.Builder userBuilder)
+            throws ActionExecutionRequestBuilderException {
+
+        Certificate certificate = preUpdatePasswordAction.getPasswordSharing().getCertificate();
+        if (isEncryptionRequired(certificate)) {
+            try {
+                X509Certificate publicCert = (X509Certificate)
+                        IdentityUtil.convertPEMEncodedContentToCertificate(certificate.getCertificateContent());
+                userBuilder.updatingCredential(getUnEncryptedCredential(), true, publicCert);
+            } catch (CertificateException e) {
+                throw new ActionExecutionRequestBuilderException("Error while building X509 certificate.", e);
+            }
+        } else {
+            userBuilder.updatingCredential(getUnEncryptedCredential(), false, null);
+        }
+    }
+
+    private Credential getUnEncryptedCredential() throws ActionExecutionRequestBuilderException {
+
+        PasswordSharing.Format passwordSharingFormat = preUpdatePasswordAction.getPasswordSharing().getFormat();
+        if (PasswordSharing.Format.SHA256_HASHED.equals(passwordSharingFormat)) {
+            Secret credentialObj;
+            try {
+                credentialObj = Secret.getSecret(userActionContext.getPassword());
+                MessageDigest digest = MessageDigest.getInstance("SHA-256");
+                byte[] byteValue = digest.digest(credentialObj.getBytes());
+                String passwordHash = Base64.encode(byteValue);
+
+                return new Credential.Builder()
+                        .type(Credential.Type.PASSWORD)
+                        .format(Credential.Format.HASH)
+                        .value(passwordHash)
+                        .algorithm(Credential.Algorithm.SHA256)
+                        .build();
+            } catch (NoSuchAlgorithmException | UnsupportedSecretTypeException e) {
+                throw new ActionExecutionRequestBuilderException("Error while hashing the credential.", e);
+            }
+
+        } else if (PasswordSharing.Format.PLAIN_TEXT.equals(passwordSharingFormat)) {
+            return new Credential.Builder()
+                    .type(Credential.Type.PASSWORD)
+                    .format(Credential.Format.PLAIN_TEXT)
+                    .value(new String(userActionContext.getPassword()))
```

Comment:
what would actually happen if we send the char array

Reply:
Use char array in the Credential object as well and see whether the deserializer can handle it

Reply:
yeah serializer and serialize the char[]. Hence we dont need to have the password as a string